tests/libvmi/cloudinit.go

```diff
@@ -79,12 +82,14 @@ func addDiskVolumeWithCloudInitConfigDrive(vmi *v1.VirtualMachineInstance, diskN
 }
 
 func addDiskVolumeWithCloudInitNoCloud(vmi *v1.VirtualMachineInstance, bus v1.DiskBus) {
-	addDisk(vmi, newDisk(cloudInitDiskName, bus))
-	v := newVolume(cloudInitDiskName)
-	setCloudInitNoCloud(&v, &v1.CloudInitNoCloudSource{})
-	addVolume(vmi, v)
-}
+	disk := newDisk(cloudInitDiskName, bus)
+	if !diskExists(vmi, disk) {
```

Comment:
The changes for disks and volumes are no longer override any existing, however they ignore the request silently.

Between ignoring the addition and overriding, I this overriding is more accurate. E.g. given a default (of a factory) we can override it.

To be clear, the only case where it is fine to ignore is when the disk/volume is exactly the same (not only the name).

Reply:
>The changes for disks and volumes are no longer override any existing, however they ignore the request silently.

I think this is ok-ish as we always use the same name for the cloud-init disk.

>Between ignoring the addition and overriding, I think overriding is more accurate. E.g. given a default (of a factory) we can override it.

I disagree because one could unintentionally override the previous version. I am ok if we make the factory options optional (see Federico's comment, I like the suggestion).

>To be clear, the only case where it is fine to ignore is when the disk/volume is exactly the same (not only the name).

I think here I should check if the volume does have the same source. It should be enough, do you agree?

Reply:
> > Between ignoring the addition and overriding, I think overriding is more accurate. E.g. given a default (of a factory) we can override it.
> 
> I disagree because one could unintentionally override the previous version. I am ok if we make the factory options optional (see Federico's comment, I like the suggestion).

This is the given behavior which this libvmi follows. Options may override previous settings and there are enough scenarios where this has been found beneficial.

Sometime tests want to be explicit and specify a specific parameter. The fact it is already set does not mean it should not be overriden. It was already set because a factory was found to prefer a specific setting, but to any preference like this we have exceptions. These exceptions are solved through overrides.

> > To be clear, the only case where it is fine to ignore is when the disk/volume is exactly the same (not only the name).
> 
> I think here I should check if the volume does have the same source. It should be enough, do you agree?

All the values need to be checked (deep-equality) and if they are not equal you are suppose to panic.
Again, this may collide with the notion of allowing overrides as discussed previously.

It will be problematic if we use sometimes panic, sometime ignore and sometime override.
IMO we should prefer override, as it requires less logic (conditions) and fits how options are treated.
But there are always exceptions, it cannot be a hard rule for sure.

Reply:
> > > Between ignoring the addition and overriding, I think overriding is more accurate. E.g. given a default (of a factory) we can override it.
> > 
> > 
> > I disagree because one could unintentionally override the previous version. I am ok if we make the factory options optional (see Federico's comment, I like the suggestion).
> 
> This is the given behavior which this libvmi follows. Options may override previous settings and there are enough scenarios where this has been found beneficial.
> 
> Sometime tests want to be explicit and specify a specific parameter. The fact it is already set does not mean it should not be overriden. It was already set because a factory was found to prefer a specific setting, but to any preference like this we have exceptions. These exceptions are solved through overrides.

I believe you are still talking about factory, where it is ok to override but we can also build other factories where overriding will be nonsense. The proposed solution by Federico would allow supplying your options and defaulting at the same time.

Overriding as a general behavior of libvmi doesn't make sense, I believe it goes against the "testing library".

> 
> > > To be clear, the only case where it is fine to ignore is when the disk/volume is exactly the same (not only the name).
> > 
> > 
> > I think here I should check if the volume does have the same source. It should be enough, do you agree?
> 
> All the values need to be checked (deep-equality) and if they are not equal you are suppose to panic. Again, this may collide with the notion of allowing overrides as discussed previously.
> 
> It will be problematic if we use sometimes panic, sometime ignore and sometime override. IMO we should prefer override, as it requires less logic (conditions) and fits how options are treated. But there are always exceptions, it cannot be a hard rule for sure.

I am proposing the opposite, prefer panic and allow explicit overrides.